src/backend/dx11/src/data.rs

```diff
@@ -206,54 +247,67 @@ pub fn map_bind(bind: Bind) -> D3D11_BIND_FLAG {
 
 pub fn map_access(access: memory::Access) -> D3D11_CPU_ACCESS_FLAG {
     let mut r = D3D11_CPU_ACCESS_FLAG(0);
-    if access.contains(memory::READ) { r = r | D3D11_CPU_ACCESS_READ }
-    if access.contains(memory::WRITE) { r = r | D3D11_CPU_ACCESS_WRITE }
+    if access.contains(memory::READ) {
+        r = r | D3D11_CPU_ACCESS_READ
+    }
+    if access.contains(memory::WRITE) {
+        r = r | D3D11_CPU_ACCESS_WRITE
+    }
     r
 }
 
-pub fn map_usage(usage: Usage) -> (D3D11_USAGE, D3D11_CPU_ACCESS_FLAG) {
+pub fn map_usage(usage: Usage) -> Result<(D3D11_USAGE, D3D11_CPU_ACCESS_FLAG), Error> {
     match usage {
-        Usage::GpuOnly => (D3D11_USAGE_DEFAULT,   D3D11_CPU_ACCESS_FLAG(0)),
-        Usage::Immutable => (D3D11_USAGE_IMMUTABLE, D3D11_CPU_ACCESS_FLAG(0)),
-        Usage::Dynamic => (D3D11_USAGE_DYNAMIC,   D3D11_CPU_ACCESS_WRITE),
-        Usage::Mappable(_access) => unimplemented!(),
-        Usage::CpuOnly(access) => (D3D11_USAGE_STAGING, map_access(access)),
+        Usage::GpuOnly => Ok((D3D11_USAGE_DEFAULT, D3D11_CPU_ACCESS_FLAG(0))),
+        Usage::Immutable => Ok((D3D11_USAGE_IMMUTABLE, D3D11_CPU_ACCESS_FLAG(0))),
+        Usage::Dynamic => Ok((D3D11_USAGE_DYNAMIC, D3D11_CPU_ACCESS_WRITE)),
+        Usage::Mappable(access) => {
+            if access.contains(memory::READ) {
+                Err(Error::Unsupported)
+            } else {
+                Ok((D3D11_USAGE_DYNAMIC, D3D11_CPU_ACCESS_WRITE))
+            }
+        }
+        Usage::CpuOnly(access) => Ok((D3D11_USAGE_STAGING, map_access(access))),
```

Comment:
I understand that you are talking about gfx's `Dynamic`, and yes it is not mappable, but you can still use `update_buffer` on it, and while it is a very limited way tu update a buffer, it does have advantages. Not enabling mapping for `Dynamic` can improve perfomances for using `update_buffer` exclusively, juste like not enabling buffer updates for `Mappable` can improve performances for mapping exclusively. Anyway, I agree that change is needed, let's continue this discussion in #970 ?

About altering parts of buffer, I think you can map buffer ranges with OpenGL but it doesn't look to be the same thing. You could open another issue for this feature ?